src/printer/tvmscript_unified_printer.h

```diff
@@ -0,0 +1,656 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * \brief Printer class to print TVMScript from Relax and TIR nodes.
+ */
+#ifndef TVM_TVMSCRIPT_UNIFIED_PRINTER_H_
+#define TVM_TVMSCRIPT_UNIFIED_PRINTER_H_
+
+#include <tvm/node/node.h>
+#include <tvm/runtime/container/map.h>
+
+#include <initializer_list>
+#include <utility>
+
+#include "tvm/relay/expr.h"
+#include "tvm/runtime/container/array.h"
+#include "tvm/runtime/object.h"
+
+namespace tvm {
+namespace script {
+
+// Code Doc
+
+class DocNode : public Object {
+ public:
+  ObjectRef origin_ir_node;
+
+  DocNode() = default;
+  virtual ~DocNode() = default;
+
+  static constexpr const char* _type_key = "script.Docs.Doc";
+  TVM_DECLARE_BASE_OBJECT_INFO(DocNode, Object);
+};
+
+class Doc : public ObjectRef {
+ public:
+  Doc() : Doc(make_object<DocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(Doc, ObjectRef, DocNode);
+};
+
+// Base Expr Doc
+
+class ExprDocNode : public DocNode {
+ public:
+  static constexpr const char* _type_key = "script.Docs.BaseExprDoc";
+  TVM_DECLARE_BASE_OBJECT_INFO(ExprDocNode, DocNode);
+};
+
+class ExprDoc : public Doc {
+ public:
+  static ExprDoc TIRBuilder();
+  static ExprDoc None();
+
+  template <typename... AttrType>
+  static ExprDoc TIRBuilderAttribute(AttrType&&... name);
+
+  ExprDoc AccessAttr(String attr);
+
+  template <typename... ArgType>
+  ExprDoc CallWith(ArgType&&... args);
+
+  template <typename... IndexType>
+  ExprDoc IndexWith(IndexType&&... args);
+
+  ExprDoc() : ExprDoc(make_object<ExprDocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(ExprDoc, Doc, ExprDocNode);
+};
+
+// Literal Value Doc
+
+class LiteralValueDocNode : public ExprDocNode {
+ public:
+  ObjectRef value;  // Can only be FloatImm/IntImm/StringImm/String
+
+  static constexpr const char* _type_key = "script.Docs.LiteralValueDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(LiteralValueDocNode, ExprDocNode);
+
+ private:
+  ObjectRef value_;
+};
+
+class LiteralValueDoc : public ExprDoc {
+ public:
+  LiteralValueDoc() : LiteralValueDoc(make_object<LiteralValueDocNode>()){};
+  LiteralValueDoc(IntImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(FloatImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(tir::StringImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(String val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
```

Comment:
remove static_cast?

```suggestion
  LiteralValueDoc(IntImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
  LiteralValueDoc(FloatImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
  LiteralValueDoc(tir::StringImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
  LiteralValueDoc(String val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
```

Reply:
This will create a delegation cycle (endless recursion into itself)